### ISSUE TITLE
combining 'run_once' with 'when' unexpected results

### ISSUE THREAD

**Issue by ghost** [ansible/ansible#11496]
Issue Type: Bug Report

Ansible Configuration: N/A

Environment: N/A

Summary:
When using run_once and when, the when clause is only evaluated on the first host in the list.

Steps To Reproduce:

```
# Assume two server, test0 and test1
- name: false on server test[0]
  stat: path=/file_that_exists_only_on_test1
  register: testfile

# This task will always be skipped, logically it _should_ run on server test[1]
- name: Run command on server the server with the file
  command: echo "It worked!"
  run_once: True
  when: testfile.stat.exists
```

Expected Results:
The last task should execute on the first matching server that matches the when clause.

Actual Results:
The last task is evaluated only on the first server in the list. Since the when condition evaluates to false on the first server, the task is 'skipped'.

**Comment by ghost** [comment]
Looking at the code, this probably isn't a Bug, and is more likely a feature request. The code explictly says "we will only send this task to the first host in the list."[1]

The only flow I can think of where this will work is if run_once is set AND a when conditional is set then walk the hosts serially evaluating the when conditional and execute on the first host the when conditional evalutes to true on.

But this may break desired behaviour in other playbooks (I can't think of an example), so maybe this works out better as a feature_request rather than a bug.

[1] https://github.com/ansible/ansible/blob/7490044bbe28029afa9e3099d86eae9fda5f88b7/lib/ansible/plugins/strategies/linear.py#L155-157

**Comment by bcoca** [comment]
run_once is designed to run only once per batch of hosts, it will pick a host, but it is not guaranteed to always be the first one in a group, it can change depending on many factors, like --limit or the host becoming unreachable. 

In your example you prevent the first host from failing (failed_when: False) so it is expected that the task runs on the first host.

closing this as it is not a bug nor a feature request.

**Comment by bcoca** [comment]
# List Information

Hi!

Thanks very much for your interest in Ansible.  It sincerely means a lot to us. 

This appears to be a user question, and we'd like to direct these kinds of things to either the mailing list or the IRC channel.
- IRC: #ansible on irc.freenode.net  
- mailing list: https://groups.google.com/forum/#!forum/ansible-project

If you can stop by there, we'd appreciate it.  This allows us to keep the issue tracker for bugs, pull requests, RFEs and the like.

Thank you once again and we look forward to seeing you on the list or IRC.  Thanks!

**Comment by ghost** [comment]
Thanks for responding bcoca, but what you said was actually incorrect.

The when conditional is checking the return code from the process, which would be 1, so the when conditional _would_ fail. 'when: status.rc == 0'

The return code on the second host would be 0, and the when conditional would succeed, but because of run_once it never even tries resulting in the task being skipped.

After examining the code further, I agree this is not a bug since it appears run_once is running as designed. But it should be at the very least a feature request.

**Comment by bcoca** [comment]
the status you are examining in the 3rd task always comes from the 2nd task, reusing the same variable name is not a good idea

**Comment by ghost** [comment]
Fair point. This was code I threw together and it is clearly not describing the issue. Let me update the code with an actual tested example that demonstrates the behavior.

Example updated in initial post.

**Comment by maurofaccenda** [comment]
having the very same issue, can we reopen this or should i create a new issue?

```
ansible-playbook 2.3.2.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Aug  8 2017, 10:46:41) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```

**Comment by promaty** [comment]
Still the same problem on Ansible 2.3.2....

**Comment by miminar** [comment]
No change on `ansible 2.4.3.0` either.

My case:

```
- debug:
    msg: "group_names: {{ group_names }}"

- name: Check if registry exists
  command: >
    oc get dc/registry -n default
    -o json
    --config=/etc/origin/master/admin.kubeconfig
  register: registry_exists
  when: "'masters' in group_names"
  run_once: true
  become: true
  changed_when: false
  failed_when: false
```

Actual result:

```
TASK [postdeploy : debug] ******************************************
ok: [ld3142.example.com] => {
    "msg": "group_names: [u'OSEv3', u'glusterfs', u'nodes']"
}
ok: [ld3143.example.com] => {
    "msg": "group_names: [u'OSEv3', u'glusterfs', u'lb', u'nodes']"
}
ok: [ld3146.example.com] => {
    "msg": "group_names: [u'OSEv3', u'glusterfs', u'nodes']"
}
ok: [ld3140.example.com] => {
    "msg": "group_names: [u'OSEv3', u'etcd', u'masters', u'nodes']"
}
ok: [ld3159.example.com] => {
    "msg": "group_names: [u'OSEv3', u'nodes']"
}
ok: [ld3160.example.com] => {
    "msg": "group_names: [u'OSEv3', u'glusterfs', u'nodes']"
}

TASK [postdeploy : Check if registry exists] ***********************
skipping: [ld3142.example.com]
```

Expected result:

`ld3140.example.com` is picked instead of `ld3142.example.com` and the task is not skipped.

Possible work-around is to use `delegate_to` instead of `when`:

```
  # when: "'masters' in group_names"
  delegate_to: "{{ groups['masters'][0] }}"
```

**Comment by alexhexabeam** [comment]
For anyone looking for a still open ticket for this, there is https://github.com/ansible/ansible/issues/23594

This ticket obviously should not have been closed.

**Comment by andygeorge** [comment]
FYI we are able to do the desired behavior using `include_tasks:`:

`driver.yml`:

```
- include_tasks: tasks_to_include.yml
  when: conditional_matching_multiple_hosts
```

`tasks_to_include.yml`:

```
- debug: msg="only run on one of the matching hosts"
  run_once: true
```

Output:

```
TASK [include_tasks] **********************************************************
Tuesday 14 August 2018  11:30:26 -0500 (0:00:02.175)       0:00:19.835 ********
skipping: [10.X.X.137]
skipping: [10.X.X.222]
skipping: [10.X.X.132]
included: tasks_to_include.yml for 10.X.X.124, 10.X.X.5, 10.X.X.50

TASK [debug] ******************************************************************
Tuesday 14 August 2018  11:30:29 -0500 (0:00:02.343)       0:00:22.178 ********
ok: [10.X.X.124] => {
    "msg": "only run on one of the matching hosts"
}
```

**Comment by alexhexabeam** [comment]
For anyone seeing this now, the fix @andygeorge has mentioned will work for Ansible 2.4+, since `include_tasks` is dynamic and actually changes the list of hosts to not include hosts that don't match the `when` clause.

For older releases, this isn't an option, since `include` keeps all hosts in the list, but skips the ones not matching the `when` clause.

**Comment by ghost** [comment]
Wow, bringing back memories. Happy to see a cleaner solution to this problem than the hacks I had to implemented back then!
